pkg/loadbalancerservice/router.go

```diff
@@ -0,0 +1,135 @@
+package loadbalancerservice
+
+import (
+	"slices"
+
+	"github.com/openshift/microshift/pkg/config"
+	"github.com/vishvananda/netlink"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+)
+
+const (
+	defaultRouterServiceName            = "router-default"
+	defaultRouterServiceNamespace       = "openshift-ingress"
+	defaultRouterServiceAnnotationKey   = "ingresscontroller.operator.openshift.io/owning-ingresscontroller"
+	defaultRouterServiceAnnotationValue = "default"
+)
+
+type serviceUpdateFunction func([]string) error
+
+func defaultRouterWatch(ipAddresses, nicNames []string, updateFunc serviceUpdateFunction, stopCh <-chan struct{}) {
+	for {
+		ips, err := defaultRouterListenAddresses(ipAddresses, nicNames)
+		if err != nil {
+			klog.Errorf("unable to determine default router listening addresses: %v", err)
+			continue
+		}
+		if err := updateFunc(ips); err != nil {
+			klog.Errorf("unable to update default router service status: %v", err)
+			continue
+		}
+		break
+	}
+
+	updateChan := make(chan netlink.AddrUpdate)
+	doneChan := make(chan struct{})
+	err := netlink.AddrSubscribe(updateChan, doneChan)
+	if err != nil {
+		klog.Errorf("unable to subscribe to IP address changes: %v", err)
+		return
+	}
+
+	for {
+		select {
+		case <-updateChan:
+			ips, err := defaultRouterListenAddresses(ipAddresses, nicNames)
+			if err != nil {
+				klog.Errorf("unable to determine default router listening addresses: %v", err)
+				break
+			}
+			err = updateFunc(ips)
+			klog.Infof("P333. This is something I am waiting for: %v", err)
+			if err != nil {
+				klog.Errorf("unable to update default router service status: %v", err)
+				break
+			}
+		case <-stopCh:
+			klog.Info("default router watcher stopping")
+			return
+		}
+	}
+}
+
+func isDefaultRouterService(svc *corev1.Service) bool {
+	annotationValue, annotationFound := svc.Labels[defaultRouterServiceAnnotationKey]
+	return annotationFound &&
+		annotationValue == defaultRouterServiceAnnotationValue &&
+		svc.Name == defaultRouterServiceName &&
+		svc.Namespace == defaultRouterServiceNamespace
+}
+
+func defaultRouterListenAddresses(ipAddresses, nicNames []string) ([]string, error) {
+	allowedAddresses, err := config.AllowedListeningIPAddresses()
+	if err != nil {
+		return nil, err
+	}
+
+	allowedNicNames, err := config.AllowedNICNames()
+	if err != nil {
+		return nil, err
+	}
+
+	ipList := make([]string, 0)
+
+	for _, ip := range ipAddresses {
+		if !slices.Contains(allowedAddresses, ip) {
+			klog.Warningf("IP address %v not found in the host. Skipping", ip)
+			continue
+		}
+		ipList = append(ipList, ip)
+	}
+
+	for _, nicName := range nicNames {
+		if !slices.Contains(allowedNicNames, nicName) {
+			klog.Warningf("NIC %v not found in the host. Skipping", nicName)
+			continue
+		}
+		nicAddresses, err := ipAddressesFromNIC(nicName)
+		if err != nil {
+			return nil, err
```

Comment:
Above and below we're continuing on "errors" but here we're quiting. What do you think if we'd continue on this error as well (or ignore it in the `defaultRouterWatch`) so this "controller" keeps running?

Reply:
When you say continuing on errors you mean skipping IPs which are not present anymore? This is not an abnormal situation in MicroShift, but in the configuration. If the values you configured are gone then it is the admin who should take action to fix it.
If, however, MicroShift is unable to retrieve host addresses or NICs then this is a different issue. Since its difficult to determine the cause from here, an error log will be written in the top handler and it will not change the last good known state. The next IP change (or an admin restarting something) will trigger this function again.
The key in this controller is avoiding changing the last good known state unless the new one is accurate.

Reply:
True, but it seems we're not communicating degraded state in any way (beside logs, but come on)? And the remediation is basically "restart microshift" - either manually or by reconfiguring some IPs?

Reply:
This shall be part of the docs. We do not have a way of signaling this kind of condition in MicroShift. Killing the process could have other implications for an app if its unexpected.

Reply:
ack